openzl-crypto/src/poseidon/constants.rs

```diff
@@ -0,0 +1,307 @@
+//! Poseidon Permutation Round Numbers
+
+use crate::poseidon::Specification;
+
+#[cfg(feature = "std")]
+use {openzl_util::num::Ceil, security::SecurityCondition};
+
+/// Poseidon Security Parameters
+///
+/// See the [`Specification`](super::Specification) for more on how these constants are part of the
+/// Poseidon Permutation algorithm.
+///
+/// # Safety
+///
+/// The constants that are specified to be "secure" in this `struct` have only been guaranteed to
+/// work for BLS12-381.
+#[derive(Clone, Copy, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct Constants {
+    /// Width of the Internal Poseidon State
+    pub width: usize,
+
+    /// Number of Full Rounds
+    pub full_rounds: usize,
+
+    /// Number of Partial Rounds
+    pub partial_rounds: usize,
+}
+
+impl Constants {
+    /// Poseidon Prime Field Modulus Bit Count
+    ///
+    /// This constant is denoted `n` in the Poseidon paper where `n = ceil(log2(p))`.
+    pub const MODULUS_BITS: usize = 255;
+
+    /// Security Level in Bits
+    ///
+    /// This constant is denoted `M` in the Poseidon paper.
+    pub const SECURITY_LEVEL: usize = 128;
+
+    /// Computes constants for a Poseidon implementation that achieves [`SECURITY_LEVEL`] bits of
+    /// security according to the claims in the Poseidon paper and current cryptanalytic efforts.
+    ///
+    /// [`SECURITY_LEVEL`]: Self::SECURITY_LEVEL
+    #[cfg(feature = "std")]
+    #[cfg_attr(doc_cfg, doc(cfg(feature = "std")))]
+    #[inline]
+    pub fn from_arity(arity: usize) -> Self {
+        let mut constants = Self {
+            width: arity + 1,
+            full_rounds: 0,
+            partial_rounds: 0,
+        };
+        let mut minimum_sbox_count = usize::MAX;
+        for mut rf in (2..=1000).step_by(2) {
+            for mut rp in 4..200 {
+                if (Self {
+                    width: constants.width,
+                    full_rounds: rf,
+                    partial_rounds: rp,
+                })
+                .are_secure()
+                {
+                    rf += 2;
+                    rp = (1.075 * rp as f32).ceil() as usize;
+                    let sbox_count = constants.width * rf + rp;
+                    if sbox_count < minimum_sbox_count
+                        || (sbox_count == minimum_sbox_count && rf < constants.full_rounds)
+                    {
+                        constants.full_rounds = rf;
+                        constants.partial_rounds = rp;
+                        minimum_sbox_count = sbox_count;
+                    }
+                }
+            }
+        }
+        constants
+    }
```

Comment:
Legitimate questions. Regarding the loop it looks like we just copied the code provided by the poseidon authors \[here\](https://extgit.iaik.tugraz.at/krypto/hadeshash/-/blob/master/code/calc_round_numbers.py). Probably no one bothered to optimize this code because you only need to run it once to figure out the parameters.  
  
I don't know how to answer the full round/partial round tradeoff question. I started doing some computations based on the number of constraints for full/partial rounds but then realized I wasn't accounting for the optimizations that combine the matrix multiplication with adding round constants, which further reduce the cost of partial rounds compared to full. It's also relevant to consider how many constraints matrix mult. costs in a given proof system -- it's more expensive to do linear combinations in plonk-like arithmetizations than R1CS.

Reply:
Moved this discussion to a new issue #42.